packages/core/src/browser/tree/tree-compression.ts

```diff
@@ -0,0 +1,122 @@
+/********************************************************************************
+ * Copyright (c) 2021 SAP SE or an SAP affiliate company and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable } from 'inversify';
+import { ExpandableTreeNode } from './tree-expansion';
+import { TreeNode, CompositeTreeNode } from './tree';
+
+export class CompressionStack {
```

Comment:
@colin-grant-work it looks much better, I'm going to try it... thank you

Reply:
> ```ts
>  export function isCompressionParent(node?: TreeNode): node is CompositeTreeNode {
>         return CompositeTreeNode.is(node) && node.children.length === 1 && CompositeTreeNode.is(node.children[0]);
>     }
> ```

@colin-grant-work where will be the check for `DirNode`? only on the `file tree`? (a single file shouldn't be compressed...)

Reply:
@colin-grant-work 
where will be all the conditions from here:
` protected shouldCompressNode(node: TreeNode): boolean {
        if (!CompositeTreeNode.is(node)) {
            return false;
        }
        const root = this.getRootNode(node);
        // The tree root is always `WorkspaceNode`, its child is the main project folder - shouldn't be compressed,
        // its grandchildren are the direct children of the main project folder - shouldn't be compressed either.
        if (node.parent && (node.parent.id === root.id || (node.parent.parent && node.parent.parent.id === root.id))) {
            return false;
        }
        return DirNode.is(node) && this.isSingleChild(node);
    }

    protected isSingleChild(node: CompositeTreeNode): boolean {
        return !!node.parent && node.parent.children.length === 1;
    }`

Reply:
@colin-grant-work one more thing is problematic - each place that uses for example `isCompressionChild` function - need to ensure that compression is enabled, otherwise - the actions will be executed also when unselect compact settings...

Reply:
> where will be the check for DirNode? only on the file tree? (a single file shouldn't be compressed...)

One option would be to put these functions not in a namespace, but on the `Tree` or `TreeWidget` itself. Then it could be overridden by the `FileTree`. But would the check for `CompositeTreeNode.is()` not cover the distinction between directories and files?

---

> where will be all the conditions from here:

In the `shouldCompressNode` function, only the `CompositeTreeNode.is(node)` refers to something that wouldn't be available with my suggested code, and that could be replaced with something like `(this.isCompressionEnabled() && CompositeTreeNode.isCompressionParticipant(node)`

---

> one more thing is problematic - each place that uses for example isCompressionChild function - need to ensure that compression is enabled, otherwise - the actions will be executed also when unselect compact settings... As I see it - the node must have the compressed property in order to indicate the enablement of the compression... I'm waiting forward to see you answer...

There are a couple of possibilities here. If the functions are left in the namespace, then they could all accept a second argument `applyCompression` (as `getCompressedItems` does in my suggestion), and only return true if that's true. Alternatively, if the functions are moved into the `Tree` system, then they can easily check `this.isCompressionEnabled`. The trickiest part is the change to the `TreeIterator` system. There, I think the most straightforward approach would be to add a third option to `TreeIterator.Options`, something like this:

```typescript
export interface Options {
        /**
         * Whether to ignore siblings of the starting node.
         */
        readonly pruneSiblings: boolean;
        /**
         * Whether to ignore nodes whose parents are not expanded.
         */
        readonly pruneCollapsed: boolean;
        /**
         * Whether to include composite nodes that are the only children of their parents
         * even if the parent is collapsed.
         * Ignored of `pruneCollapsed` is `false`.
         */
        readonly applyCompression: boolean;
    }
```

The general question is whether we want `CompressibleTreeNode` to be a new entity, or whether we want compression to be a new type of tree behavior that depends on other state / flags. I think it's predictable enough to be a form of behavior, but if it turns out that we need more entities to track the state accurately.

Reply:
@colin-grant-work  thank you for the clear answers, 
> The trickiest part is the change to the `TreeIterator` system

Beside the `TreeIterator` - the `TreeExpansionServiceImpl` also uses the compression... 
I'm able to put all the compression functions in the `CompressibleTree` and check in `TreeExpansionServiceImpl`: ``if('isCompressionParticipiant' in this.tree){return this.tree as any.isCompressionParticipiant}`` ...

Should I do this or leave all and stay with `CompressibleTreeNode` (without the `compressibleTreeService`), Pay attention that it might be that in the future will be more places that need to use it (e.g in `selectionTree` creating `TreeIterator` without the compression childs because the `applyCompression` info won't be available to him)